tools/spin-ec2/main.go

```diff
@@ -147,3 +167,108 @@ func main() {
 		}
 	}
 }
+
+func cliMode() {
+
+	var cmdCanaries = &cobra.Command{
+		Use:   "canaries",
+		Short: "Canary machines tools for infrastructure-agent",
+		Long:  `canaries command is used for infrastructure-agent canary machines.`,
+		Args:  cobra.MinimumNArgs(1),
+		Run:   func(cmd *cobra.Command, args []string) {},
+	}
+
+	var cmdProvision = &cobra.Command{
+		Use:   "provision",
+		Short: "Provision canary machines",
+		Long:  `provision is used to deploy canary machines with infrastructure-agent installed.`,
+		RunE:  provisionCanaries,
+	}
+
+	// Infra agent version to install.
+	cmdProvision.PersistentFlags().StringP("agent_version", "v", "", "infrastructure-agent version to deploy")
+	viper.BindPFlag("agent_version", cmdProvision.PersistentFlags().Lookup("agent_version"))
+	cmdProvision.MarkPersistentFlagRequired("agent_version")
+
+	// NR license key.
+	cmdProvision.PersistentFlags().StringP("license", "l", "", "infrastructure-agent license key")
+	viper.BindPFlag("license", cmdProvision.PersistentFlags().Lookup("license"))
+	cmdProvision.MarkPersistentFlagRequired("license")
+
+	var cmdPrune = &cobra.Command{
+		Use:   "prune",
+		Short: "Prune canary machines",
+		Long:  `prun is used to remove old canary machines.`,
+		RunE:  pruneCanaries,
+	}
+
+	cmdPrune.PersistentFlags().Bool("dry_run", false, "dry run")
+	viper.BindPFlag("dry_run", cmdPrune.PersistentFlags().Lookup("dry_run"))
+
+	cmdRoot := &cobra.Command{Use: "spin-ec2"}
+	cmdRoot.AddCommand(cmdCanaries)
+	cmdCanaries.AddCommand(cmdProvision, cmdPrune)
+	cmdRoot.Execute()
+}
+
+// provisionCanaries will provision aws machines with the infra-agent installed.
+func provisionCanaries(cmd *cobra.Command, args []string) error {
+	agentVersion := viper.GetString("agent_version")
+	license := viper.GetString("license")
+
+	if !semver.IsValid(agentVersion) {
+		return fmt.Errorf("agent version '%s' doesn't match the pattern 'vmajor.minor.patch' format",
+			agentVersion)
+	}
+
+	ansibleGroupVars, err := readAnsibleGroupVars()
+	if err != nil {
+		return err
+	}
+
+	opts, err := generateOptions(*ansibleGroupVars)
+	if err != nil {
+		return err
+	}
+
+	prepareAnsibleConfig(opts, fmt.Sprintf("%s:%s", hostPrefix, agentVersion))
+
+	curPath, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	execNameArgs("ansible-playbook",
+		"-i", path.Join(curPath, "test/automated/ansible/inventory.local"),
+		"--extra-vars", "@"+path.Join(curPath, inventory),
+		path.Join(curPath, "test/automated/ansible/provision.yml"))
+
+	provisionOpts := newProvisionOptions()[6]
```

Comment:
Probably we need to export each option individually and import by name here. It is not reliable to relay on index in this case.